### PR TITLE
feat: validate withdrawal when validating block.

### LINKED ISSUE #1399: Validate that withdrawal index is monotonically increasing
Ref https://github.com/paradigmxyz/reth/pull/1322#discussion_r1108319581

Validate that the index of the first withdrawal is monotonically increasing in relation to the last processed withdrawal (in the previous block or earlier).

This check should be added in `validate_block_regarding_chain` function
https://github.com/paradigmxyz/reth/blob/main/crates/consensus/src/validation.rs#L295-L302

It also requires adding a `WithdrawalProvider` in order to look up the last processed withdrawal.

### PR DESCRIPTION
**Pull request by GrapeBaBa**
Fix #1399

### CODE REVIEW

**Comment by GrapeBaBa** [comment]
@rakita Is it expected? I will add test for this feature.

File: crates/storage/provider/src/traits/withdrawals.rs [hunk 1]
```diff
@@ -5,4 +5,7 @@ use reth_primitives::{BlockId, Withdrawal};
 pub trait WithdrawalsProvider: Send + Sync {
     /// Get withdrawals by block id.
     fn withdrawals_by_block(&self, id: BlockId, timestamp: u64) -> Result<Option<Vec<Withdrawal>>>;
+
+    /// Get lastest withdrawal from this block or earlier .
+    fn lastest_withdrawal(&self, id: BlockId, timestamp: u64) -> Result<Option<Withdrawal>>;
```
Comment (rkrasiuk) [inline crates/storage/provider/src/traits/withdrawals.rs:10]:
```suggestion
    fn latest_withdrawal(&self, id: BlockId, timestamp: u64) -> Result<Option<Withdrawal>>;
```

File: crates/consensus/src/validation.rs [hunk 2]
```diff
@@ -353,12 +354,25 @@ pub fn validate_block_regarding_chain<PROV: HeaderProvider>(
...
         .ok_or(Error::ParentUnknown { hash: block.parent_hash })?;
 
+    // Check if withdrawals are valid.
+    let parent_block_id = BlockId::Number(BlockNumberOrTag::Number(parent.number));
+    let latest_withdrawal = provider.lastest_withdrawal(parent_block_id, parent.timestamp)?;
+    if let Some(withdrawals) = &block.withdrawals {
+        if !withdrawals.is_empty() {
+            let withdrawal_index = withdrawals.first().unwrap().index;
+            let lastest_withdrawal_index = latest_withdrawal.map_or(0, |w| w.index);
+            if lastest_withdrawal_index + 1 != withdrawal_index {
```
Comment (rkrasiuk) [inline crates/consensus/src/validation.rs:364]: the first withdrawal should have index 0. this logic would fail, because it expects index 1 if no latest withdrawal was found

File: crates/consensus/src/validation.rs [hunk 3]
```diff
@@ -353,12 +354,25 @@ pub fn validate_block_regarding_chain<PROV: HeaderProvider>(
...
+                return Err(Error::BlockKnown { hash, number: block.header.number }.into())
```
Comment (rkrasiuk) [inline crates/consensus/src/validation.rs:365]: create a proper error variant

File: crates/storage/provider/src/providers/mod.rs [hunk 4]
```diff
@@ -218,6 +218,25 @@ impl<DB: Database> WithdrawalsProvider for ShareableDatabase<DB> {
...
+            None => Ok(None),
+            Some(withdrawals) => {
+                if !withdrawals.is_empty() {
+                    Ok(withdrawals.last().cloned())
+                } else {
+                    // find the lastest withdrawal in previous block
+                    let number = self.block_number_for_id(id)?.unwrap();
+                    if number == 0 {
+                        return Ok(None)
+                    }
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/mod.rs:233]: instead of this check if the timestamp is pre shanghai

File: crates/storage/provider/src/providers/mod.rs [hunk 5]
```diff
@@ -218,6 +218,25 @@ impl<DB: Database> WithdrawalsProvider for ShareableDatabase<DB> {
...
+                    let id = BlockId::Number((number - 1).into());
+                    self.lastest_withdrawal(id, timestamp)
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/mod.rs:235]: this function is doing a lot of redundant database lookups. instead of recursion, create a cursor on withdrawals table. it does not store empty withdrawals, so the previous entry in the database would be the value you are looking for

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/storage/provider/src/providers/mod.rs (hunk 5 above)
Comment (GrapeBaBa) [reply crates/storage/provider/src/providers/mod.rs:235]: yes, I also think this way is not good, let me look at the code you mentioned.

File: crates/consensus/src/validation.rs [hunk 6]
```diff
@@ -471,6 +499,22 @@ mod tests {
         }
     }
 
+    impl WithdrawalsProvider for Provider {
+        fn latest_withdrawal(
```
Comment (GrapeBaBa) [inline crates/consensus/src/validation.rs:503]: for test, the current code like this sees can only mock one kind of case? I need test both None and Some, any existing approach in reth codebase?

**Comment by GrapeBaBa** [comment]
wait for #1609

**Comment by gakonst** [comment]
@rkrasiuk is this a blocker / needs to be closed for Shanghai? or just a "nice to have"? if it's needed, @GrapeBaBa please let us know if you'll get to it, else would like to assign to someone else

**Comment by GrapeBaBa** [comment]
> @rkrasiuk is this a blocker / needs to be closed for Shanghai? or just a "nice to have"? if it's needed, @GrapeBaBa please let us know if you'll get to it, else would like to assign to someone else

Sorry for late, already introduce mockall in this PR to fix #1609

File: crates/storage/provider/src/providers/mod.rs [hunk 7]
```diff
@@ -267,6 +267,17 @@ impl<DB: Database> WithdrawalsProvider for ShareableDatabase<DB> {
         }
         Ok(None)
     }
+
+    fn latest_withdrawal(&self) -> Result<Option<Withdrawal>> {
```
Comment (gakonst) [inline crates/storage/provider/src/providers/mod.rs:271]: We can add a test for this in a follow-up PR, is a simple enough change

File: crates/consensus/src/validation.rs [hunk 8]
```diff
@@ -656,6 +724,39 @@ mod tests {
             validate_block_standalone(&block, &chain_spec),
             Err(ConsensusError::WithdrawalIndexInvalid { .. })
         );
+
+        let (_, parent) = mock_block();
+        let mut provider = Provider::new(Some(parent.clone()));
+        // Withdrawal index should be 0 if there are no withdrawals in the chain
+        let block = create_block_with_withdrawals(&[1, 2, 3]);
+        provider.withdrawals_provider.expect_latest_withdrawal().return_const(Ok(None));
```
Comment (gakonst) [inline crates/consensus/src/validation.rs:732]: pretty nice API :) I think we've done this manually in a lot of other places, maybe worth investigating as an alternative in our testing, cc @onbjerg @mattsse @rkrasiuk

**Review by gakonst — APPROVED** [review]
